provider/equinix/environ.go

```diff
@@ -0,0 +1,784 @@
+	matchingTools, err := args.Tools.Match(tools.Filter{Arch: spec.Image.Arch})
+	if err != nil {
+		return errors.Errorf("chosen architecture %v for image %q not present in %v",
+			spec.Image.Arch, spec.Image.Id, args.Tools.Arches())
+	}
+
+	if spec.InstanceType.Deprecated {
+		logger.Warningf("deprecated instance type specified: %s", spec.InstanceType.Name)
+	}
+
+	if err := args.InstanceConfig.SetTools(matchingTools); err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := instancecfg.FinishInstanceConfig(args.InstanceConfig, e.Config()); err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+var ErrDeviceProvisioningFailed = errors.New("device provisioning failed")
+
+// waitDeviceActive is a function capable of figuring out when a Equinix Metal
+// device is active
+func waitDeviceActive(ctx context.ProviderCallContext, c *packngo.Client, id string) (*packngo.Device, error) {
+	var d *packngo.Device
+	err := retry.Call(retry.CallArgs{
+		Func: func() error {
+			var err error
+			d, _, err = c.Devices.Get(id, nil)
+			if err != nil {
+				return err
+			}
+			if d.State == "active" {
+				return nil
+			}
+			if d.State == "failed" {
+				return errr.Wrap(ErrDeviceProvisioningFailed, fmt.Sprintf("device %s provisioning failed", id))
+			}
+			return fmt.Errorf("device in not in active state yet")
+		},
+		IsFatalError: func(err error) bool {
+			if errr.Is(err, ErrDeviceProvisioningFailed) {
+				return true
+			}
+			select {
+			case <-ctx.Dying():
+				return true // treat error as fatal because we are asked to abort
+			default:
+				return common.IsCredentialNotValid(err)
+			}
+		},
+		Attempts: 180,
+		Delay:    5 * time.Second,
+		Clock:    clock.WallClock,
+	})
+
+	if d != nil {
+		return d, nil
+	}
+
+	return nil, err
+}
+
+// Helper function to get supported OS version
+func isDistroSupported(os packngo.OS, ic *instances.InstanceConstraint) bool {
+	switch os.Distro {
+	case "ubuntu":
+		series, err := series.VersionSeries(os.Version)
+		if err != nil || ic.Series != series {
+			return false
+		}
+	case "centos":
+		series, err := series.CentOSVersionSeries(os.Version)
+		if err != nil || ic.Series != series {
+			return false
+		}
+	case "windows":
+		series, err := series.WindowsVersionSeries(os.Version)
+		if err != nil || ic.Series != series {
+			return false
+		}
+	default:
+		return false
+	}
+
+	return true
+}
+
+// helper function which tries to extract processor architecture from plan name.
+// plan names have format like c2.small.arm where in majority of cases the last bit indicates processor architecture.
+// in some cases baremeta_1 and similar are returned which are mapped to AMD64.
+func getArchitectureFromPlan(p string) string {
+	planSplit := strings.Split(p, ".")
+	var architecture string
+	if len(planSplit) > 2 {
+		architecture = planSplit[2]
+	}
+	switch architecture {
+	case "x86":
+		return arch.AMD64
+	case "arm":
+		return arch.ARM64
+	default:
+		return arch.AMD64
```

Comment:
Should this not error out, to prevent users from getting an AMD64 by accident or at least log a warning explaining the fallback to AMD64.

Reply:
@displague what do you think about this? @achilleasa